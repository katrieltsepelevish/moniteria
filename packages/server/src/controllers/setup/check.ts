import { NextFunction, Request, Response } from 'express';

import Setup from '../../models/setup';

export default async (req: Request, res: Response, next: NextFunction) => {
  try {
    const count = await Setup.countDocuments();

    const hasSetup = count > 0;

    return res.status(200).send({ completed: hasSetup });
  } catch (err) {
    next(err);
  }
};
