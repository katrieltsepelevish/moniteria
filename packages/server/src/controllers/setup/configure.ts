import { NextFunction, Request, Response } from 'express';

import Setup from '../../models/setup';
import HttpException from '../../exceptions/http';

export default async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { completed } = req.body;

    const count = await Setup.countDocuments();

    const hasSetup = count > 0;

    if (hasSetup) throw new HttpException(500, 'Setup already configured.');

    const setup = await Setup.create({
      completed,
    });

    return res.status(200).send({ setup });
  } catch (err) {
    next(err);
  }
};
