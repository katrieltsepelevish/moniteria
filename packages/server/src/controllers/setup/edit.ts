import { NextFunction, Request, Response } from 'express';

import Setup from '../../models/setup';

export default async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { configured } = req.body;

    const setup = await Setup.findOne();

    setup!.completed = configured;

    await setup!.save();

    return res.status(200).send({ setup });
  } catch (err) {
    next(err);
  }
};
