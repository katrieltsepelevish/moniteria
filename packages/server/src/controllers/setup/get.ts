import { NextFunction, Request, Response } from 'express';

import Setup from '../../models/setup';

export default async (req: Request, res: Response, next: NextFunction) => {
  try {
    const setup = await Setup.findOne();

    return res.status(200).send({ setup });
  } catch (err) {
    next(err);
  }
};
