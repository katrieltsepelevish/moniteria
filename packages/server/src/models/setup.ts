import Mongoose from 'mongoose';

const SetupSchema = new Mongoose.Schema(
  {
    completed: {
      type: Boolean,
      required: true,
    },
  },
  { timestamps: true }
);

export interface SetupDocument extends Mongoose.Document {
  completed: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export default Mongoose.model<SetupDocument, Mongoose.Model<SetupDocument>>(
  'Setup',
  SetupSchema
);
