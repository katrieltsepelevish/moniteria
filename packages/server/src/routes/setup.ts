import Express from 'express';

import checkController from '../controllers/setup/check';
import configureController from '../controllers/setup/configure';
import editController from '../controllers/setup/edit';
import getController from '../controllers/setup/get';

const router = Express.Router();

router.get('/check', checkController);
router.get('/', getController);
router.post('/', configureController);
router.patch('/', editController);

export default router;
