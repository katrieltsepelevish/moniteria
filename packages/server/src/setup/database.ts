import Mongoose, { Mongoose as MongooseInstance } from 'mongoose';

import config from '../config';
import logger from '../logger';

export let mongoClient: MongooseInstance;

export default async () => {
  Mongoose.set('strictQuery', false);

  mongoClient = await Mongoose.connect(config.databaseUri);

  logger.info('Trying to connect MongoDB...');

  logger.info('💾 Connected to MongoDB');

  mongoClient.connection.on('disconnected', () => {
    logger.info('🔌 Disconnected from MongoDB');
  });

  mongoClient.connection.on('error', (err: any) => {
    logger.error('❌ MongoDB failed', {
      message: err?.message ?? err,
      trace: err?.trace,
    });
  });

  mongoClient.connection.on('SIGINT', () => {
    mongoClient.connection.close(() => {
      logger.info('💀 MongoDB connection closed through app terminal');
      process.exit(1);
    });
  });
};
