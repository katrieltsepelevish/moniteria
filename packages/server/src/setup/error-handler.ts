import { Express, NextFunction, Request, Response } from 'express';

export default (app: Express) => {
  // Catch 404 errors and forward to error handler
  app.use(async (err: any, req: Request, res: Response, next: NextFunction) => {
    return res.status(err.status || 500).json({ message: err.message });
  });
};
