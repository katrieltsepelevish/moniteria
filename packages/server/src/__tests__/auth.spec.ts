import SuperTest, { Test } from 'supertest';

import User from '../models/user';
import jestServer from '../jestServer';
import { mongoClient } from '../setup/database';

const setup = async () => SuperTest(await jestServer());

let request: SuperTest.SuperTest<Test>;

beforeAll(async () => {
  request = await setup();
});

afterAll(async () => await mongoClient.connection.close());

describe('AUTH API ENDPOINTS', () => {
  describe('POST LOGIN /api/v1/auth/login', () => {
    it('Should not login user without passing data', async () => {
      const { status, body } = await request
        .post('/api/v1/auth/login')
        .send({});

      expect(status).toBe(422);
      expect(body).toMatchSnapshot();
    });

    it('Should successfully login a user', async () => {
      const testUser = {
        name: 'Moniteria',
        email: 'hi@moniteria.com',
        password: '2023',
      };

      await request.post('/api/v1/auth/register').send(testUser);

      const { status, body } = await request.post('/api/v1/auth/login').send({
        email: testUser.email,
        password: testUser.password,
      });

      expect(status).toBe(200);
      // expect(body.token).toBeTruthy();

      // const verifiedUser = await User.verifyToken(body.token);

      expect(body.user.email).toBe(testUser?.email);

      await User.findOneAndRemove({
        email: testUser.email,
      });
    });
  });

  describe('POST REGISTER /api/v1/auth/register', () => {
    it('Register without passing data should fail', async () => {
      const { status, body } = await request
        .post('/api/v1/auth/register')
        .send({});

      expect(status).toBe(422);
      expect(body).toMatchSnapshot();
    });

    it('Should successfully register a new user', async () => {
      const testUser = {
        name: 'Moniteria',
        email: 'hi@moniteria.com',
        password: '2023',
      };

      const { status, body } = await request
        .post('/api/v1/auth/register')
        .send(testUser);

      expect(status).toBe(201);
      // expect(body.token).toBeTruthy();

      const user = await User.findOne({
        email: testUser.email,
      });

      expect(user?.email).toBe(testUser.email);
      expect(user?.name).toBe(testUser.name);

      // const verifiedUser = await User.verifyToken(body.token);

      expect(user?.email).toBe(testUser?.email);

      await User.findOneAndRemove({
        email: testUser.email,
      });
    });
  });
});
