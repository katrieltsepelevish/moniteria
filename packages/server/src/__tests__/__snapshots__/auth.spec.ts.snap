// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AUTH API ENDPOINTS POST LOGIN /api/v1/auth/login Should not login user without passing data 1`] = `
{
  "errors": {
    "email": "email is required",
    "password": "password is required",
  },
  "message": "Validation failed.",
}
`;

exports[`AUTH API ENDPOINTS POST REGISTER /api/v1/auth/register Register without passing data should fail 1`] = `
{
  "errors": {
    "email": "email is required",
    "name": "name is required",
    "password": "password is required",
  },
  "message": "Validation failed.",
}
`;
